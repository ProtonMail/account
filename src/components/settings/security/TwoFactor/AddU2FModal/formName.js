import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';

import styles from './index.css';
import settingsActions from '../../../../../actions/settings';
import { connect } from 'unistore/full/preact';

/**
 * Label form for the U2F key.
 */
const FormName = ({ onSubmit, onCancel, addU2FKeyNameAction, settings: { addU2FKey: addU2FKeyStore } }) => {
    const model = { name: (addU2FKeyStore.response ? addU2FKeyStore.response.name : '') };
    return (<ModalWrapper
        onSubmit={(e) => {
            e.preventDefault();
            addU2FKeyNameAction(model.name);
            onSubmit();
        }}
        onReset={(e) => {
            e.preventDefault();
            onCancel();
        }}>
        <ModalContent>
            <div class={styles.name}>
                <label class={styles.nameLabel} htmlFor="name">
                    Name
                </label>
                <div class={styles.nameInputContainer}>
                    <input
                        onInput={({ target: { value } }) => {
                            model.name = value;
                        }}
                        value={model.name}
                        type="name"
                        id="inputName"
                        required
                        placeholder="Name"
                        class={styles.nameTextInput}
                        ref={input => input && input.focus()}
                    />
                </div>
            </div>
        </ModalContent>
        <ModalFooter>
            <button type="reset" value="Reset">
                Back
            </button>
            <button type="submit" value="Submit">
                Next
            </button>
        </ModalFooter>
    </ModalWrapper>);
};

export default connect('settings', settingsActions)(FormName);
