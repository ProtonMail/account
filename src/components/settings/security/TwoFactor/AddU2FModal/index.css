.name {
    display: flex;
    justify-content: space-between;
    align-self: stretch;
}

.name .label {
    flex: 1;
}

.name .inputContainer {
    flex: 4;
}

.name .textInput {
    width: 100%
}

.RegisterKeyForm-container {
    align-items: stretch;
}

.RegisterKeyForm-status {
    margin: auto 20%;
}

.RegisterKeyForm-status > div {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.RegisterKeyForm-status > div > span {
    flex: 1;
}

.RegisterKeyForm-status > div > span:first-child {
    text-align: left;
}

span.label {
    color: #9592d1;
}
