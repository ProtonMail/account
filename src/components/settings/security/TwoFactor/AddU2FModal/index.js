import Presentation from './presentation';
import FormName from './formName';
import RegisterKeyForm from './formRegisterKey';
import { steps as RecoveryCodeSteps, beforeDismiss as beforeDismissRecoveryCodeModal } from '../SaveRecoveryCodeModal';
import { steps as ScopeModal, beforeDismiss as beforeDismissScopeModal } from '../../../../auth/ScopeModal/index';
import store from '../../../../../helpers/store';
import actions from '../../../../../actions/settings';

export const steps = [
    {
        title: 'Register new U2F Key',
        component: ({ onNextStep, onPreviousStep, message }) => (
            <Presentation onSubmit={onNextStep} onCancel={onPreviousStep} message={message} />
        )
    },
    ...ScopeModal('password'),
    {
        title: 'Name your U2F Key',
        component: ({ onNextStep, onPreviousStep }) => <FormName onSubmit={onNextStep} onCancel={onPreviousStep} />
    },
    {
        title: 'Register new U2F Key',
        component: ({ onNextStep, onPreviousStep, forbidClosure, onReset }) => (
            <RegisterKeyForm
                onSubmit={onNextStep}
                onCancel={onPreviousStep}
                forbidClosure={forbidClosure}
                onReset={onReset}
            />
        )
    },
    ...RecoveryCodeSteps
];

export const beforeDismiss = () => {
    beforeDismissScopeModal();
    beforeDismissRecoveryCodeModal();
    const { resetStoreAction } = actions(store);
    resetStoreAction(store.getState(), ['addU2FKey']);
};
