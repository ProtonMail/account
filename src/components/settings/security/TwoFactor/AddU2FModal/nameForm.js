import { h, Component } from 'preact';
import { styles as ModalStyles } from '../../../../ui/Modal';
import styles from './index.css';

/**
 * Label form for the U2F key.
 */
export default class NameForm extends Component {

    /**
     * triggered each time the name is updated.
     * @param event
     */
    onNameUpdated ( event ) {
        this.setState({ name: event.target.value });
    }

    /**
     * Submit the form.
     * @param e
     */
    onSubmit ( e ) {
        e.preventDefault();
        this.props.onSubmit({ name: this.state.name });
    }

    /**
     * @constructor
     */
    constructor ( props ) {
        super(props);
        this.state = {
            name: ''
        };
    }

    render () {
        return (
            <form
                className={ModalStyles.wrapper}
                onSubmit={this.onSubmit.bind(this)}
                onReset={( e ) => {
                    e.preventDefault();
                    this.props.onCancel({});
                }}
            >
                <div className={ModalStyles.content}>
                    <div class={styles[ 'NameForm-name' ]}>
                        <label htmlFor="name">
                            Name
                        </label>
                        <div>
                            <input
                                onInput={this.onNameUpdated.bind(this)}
                                value={this.state.name}
                                type="name"
                                id="inputName"
                                placeholder="Name"
                            />
                        </div>
                    </div>
                </div>
                <div class={ModalStyles.footer}>
                    <button type="reset" value="Reset">
                        Back
                    </button>
                    <button type="submit" disabled={!this.state.name} value="Submit">
                        Next
                    </button>
                </div>
            </form>
        );
    }
}
