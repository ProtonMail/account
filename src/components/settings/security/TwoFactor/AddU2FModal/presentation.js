import { h } from 'preact';

import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';
import Link from '../../../../ui/Link';

/**
 * Shows information about the U2F Key usage.
 * @param {Object} props
 * @param {Function} props.onSubmit - triggers the next step.
 * @param {Function} props.onCancel - triggers the previous step.
 * @return {*}
 */
const Presentation = ({ onSubmit, onCancel, message }) => (
    <ModalWrapper
        onSubmit={(e) => {
            e.preventDefault();
            onSubmit();
        }}
        onReset={(e) => {
            e.preventDefault();
            onCancel();
        }}
    >
        {message ? (
            <ModalContent>
                <div>{message}</div>
            </ModalContent>
        ) : (
            <ModalContent>
                <p>This wizard will add a new security key to your Proton Account. </p>
                <p>
                    Please note that you will not be able to access your account if you loose your U2F device and your
                    recovery codes. We recommend setting up a second 2FA method as a backup.
                </p>
                <p>
                    <strong>
                        If you have never used 2FA before, we strongly recommend you reading our 2FA Guide first.
                    </strong>
                </p>
                <div>
                    <Link href="https://protonmail.com/blog/">READ U2F GUIDE </Link>
                </div>
            </ModalContent>
        )}
        <ModalFooter>
            <button type="reset" value="Reset">
                Back
            </button>
            <button type="submit" value="Submit">
                Next
            </button>
        </ModalFooter>
    </ModalWrapper>
);

export default Presentation;
