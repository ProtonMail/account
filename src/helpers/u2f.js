import { sign, register } from 'u2f-api';
import appProvider from 'frontend-commons/src/appProvider';

export const ERROR_CODE = {
    SUCCESS: 0,
    OTHER_ERROR: 1,
    BAD_REQUEST: 2,
    CONFIGURATION_UNSUPPORTED: 3,
    DEVICE_INELIGIBLE: 4,
    TIMEOUT: 5
};

const ERROR_MAP = {
    [ERROR_CODE.SUCCESS]: () => {
        throw new Error('This is a success!');
    },
    [ERROR_CODE.OTHER_ERROR]: () => 'An error occurred',
    [ERROR_CODE.BAD_REQUEST]: () => 'An internal error occurred.',
    [ERROR_CODE.CONFIGURATION_UNSUPPORTED]: () => 'This security key is not supported.',
    [ERROR_CODE.DEVICE_INELIGIBLE]: (register) =>
        register ? 'This security key is already registered for your account!' : 'This security key is not recognized.',
    [ERROR_CODE.TIMEOUT]: () =>
        'Looks like you are taking too long to respond, please try again with a bit of motivation.'
};

/**
 * Sends a SIGN request to the U2F device
 * @param {Object} U2FRequest - the U2FRequest
 * @param {Object[]} U2FRequest.RegisteredKeys - the registered keys
 * @param {String} U2FRequest.Challenge - the challenge
 * @return {Promise<SignResponse>}
 */
export async function signU2F({ RegisteredKeys: registeredKeys, Challenge: challenge }) {
    const { appId, timeout } = appProvider.getConfig('u2f');

    const signRequest = registeredKeys.map(({ Version: version, KeyHandle: keyHandle }) => ({
        version,
        keyHandle,
        appId,
        challenge
    }));

    return await sign(signRequest, timeout);
}

/**
 * Sends a REGISTER request to the U2F device.
 * @param {Object} U2FRequest - the U2FRequest
 * @param {Object[]} U2FRequest.RegisteredKeys - the registered keys
 * @param {String} U2FRequest.Challenge - the challenge
 * @param {String[]} U2FRequest.Versions - the different versions accepted by the server
 * @return {Promise<RegisterResponse>}
 */
export async function registerU2F({ RegisteredKeys: registeredKeys, Challenge: challenge, Versions: versions }) {
    const { appId, timeout } = appProvider.getConfig('u2f');
    const signRequest = registeredKeys.map(({ Version: version, KeyHandle: keyHandle }) => ({
        version,
        keyHandle
    }));

    const registerRequests = versions.map((version) => ({
        version,
        challenge,
        appId
    }));

    return await register(registerRequests, signRequest, timeout);
}

/**
 * Computes an error message from a U2F error code.
 * @param {number} errorCode - an error code.
 * @param register
 * @return {*}
 */
export function getErrorMessage(errorCode, register = false) {
    if (ERROR_MAP[errorCode]) {
        return ERROR_MAP[errorCode](register);
    }
    throw new Error(`The error code "${errorCode}" does not exist`);
}
