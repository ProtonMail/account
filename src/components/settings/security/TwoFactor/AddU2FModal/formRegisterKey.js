import { Component } from 'preact';
import { connect } from 'unistore/full/preact';
import settingsActions from '../../../../../actions/settings';

import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';
import TextButton from '../../../../ui/TextButton';

import styles from './index.css';

import image from './sign-u2f.png';
import { getErrorMessage } from '../../../../../helpers/u2f';

/**
 * Modal Form to register a new U2F Key.
 *
 * Fetches the challenge from the server, forward it to U2F API, and sends back the answer to the API.
 */
export class FormRegisterKey extends Component {

    /**
     * when next button is pressed.
     * @param {Event} e
     */
    onSubmit(e) {
        e.preventDefault();
        this.props.onSubmit();
    }

    componentDidMount() {
        this.props.addU2FKeyRegisterAction();
    }

    componentWillReceiveProps(newProps) {
        const {
            settings: { addU2FKey: { status: newStatus } }
        } = newProps;

        const {
            settings: { addU2FKey: { status } }
        } = this.props;

        if (newStatus !== status && newStatus === 'finished') {
            this.props.forbidClosure();
        }
    }

    /**
     * Renders the status field.
     * @return {*}
     */
    renderStatus() {
        const {
            settings: { addU2FKey: { response: { name } = {}, status, request, error } }
        } = this.props;

        if (status !== 'failure') {
            return (<div class={styles.status}>
                <div class={styles.row}>
                    <span class={styles.text}>Activate your key</span>
                    <span class={styles.text}>
                            {status || 'fetching'}...
                        </span>
                </div>

                <div class={styles.row}>
                    <span class={styles.text}>Name</span>
                    <span class={[styles.text, styles.nameLabel].join(' ')}>{name}</span>
                </div>
            </div>);

        }
        const { metaData: { code } = {} } = error;
        if (code) {
            return (<div class={styles.status}>
                <span>{getErrorMessage(code, true)}</span>
                <TextButton onClick={() => this.props.addU2FKeyRegisterAction()}>Retry</TextButton>
            </div>);
        }

        this.props.onReset(error.message);
    }

    render () {
        const {
            settings: { addU2FKey: { status } }
        } = this.props;

        return (
            <ModalWrapper
                onSubmit={(e) => this.onSubmit(e)}
                onReset={(e) => {
                    e.preventDefault();
                    this.props.onCancel();
                }}
            >
                <ModalContent class={styles.container}>
                    <img src={image}/>

                    {this.renderStatus()}
                </ModalContent>
                <ModalFooter>
                    <button type="reset" value="Reset" disabled={(status === 'finished')}>
                        Back
                    </button>
                    <button type="submit" value="Submit" disabled={(status !== 'finished')}>
                        Next
                    </button>
                </ModalFooter>
            </ModalWrapper>
        );
    }
}


export default connect(['scope', 'settings'], settingsActions)(FormRegisterKey);
