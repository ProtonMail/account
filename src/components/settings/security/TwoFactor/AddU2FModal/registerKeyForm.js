import { h, Component } from 'preact';
import { register } from 'u2f';

import { styles as ModalStyles } from '../../../../ui/Modal';
import styles from './index.css';

import image from './sign-u2f.png';
import appProvider from 'frontend-commons/src/appProvider';

export default class RegisterKeyModal extends Component {
    onRegisterSucceed ( result ) {
        this.setState({ result, status: 'success' });
    }

    onRegisterFails ( result ) {
        this.setState({ result, status: 'failure' });
    }

    onSubmit ( e ) {
        e.preventDefault();
        this.props.onSubmit({
            codes: [
                'd1e4822e',
                '2d83d85b',
                '717ebed4',
                'a9ffee38',
                'f34ebdf7',
                'd1321481',
                '2572f6cd',
                '39bd63b3',
                'c8016641',
                'a9b4cf9d',
                '8f475f77',
                '123d33b1'
            ]
        });
    }

    constructor () {
        super();
        this.state = {
            status: 'initialize',
            request: {
                Challenge: 'KGOcAJhmPwMKZ4r8vFb2vZZktLh9wZCmLKHxQQH1bxY',
                Versions: [ 'U2F_V2' ],
                RegisteredKeys: []
            }
        };
    }

    componentWillMount () {
        if (this.state.status === 'initialize') {
            const u2fConfig = appProvider.getConfig('u2f');
            register(this.state.request, u2fConfig.appID, u2fConfig.timeout)
                .then(this.onRegisterSucceed.bind(this))
                .catch(this.onRegisterFails.bind(this));
            this.setState({ status: 'waiting' });
        }
    }

    render () {
        const name = this.props.params.name;
        return (
            <form
                class={ModalStyles.wrapper}
                onSubmit={this.onSubmit.bind(this)}
                onReset={( e ) => {
                    e.preventDefault();
                    this.props.onCancel({});
                }}
            >
                <div class={[ styles[ 'RegisterKeyForm-container' ], ModalStyles.content ].join(' ')}>
                    <img src={image}/>

                    <div class={styles[ 'RegisterKeyForm-status' ]}>
                        <div>
                            <span>Activate your key</span>
                            <span>
                                {this.state.status === 'initialize' ? 'fetching' : this.state.status}...
                            </span>
                        </div>

                        <div>
                            <span>Name</span>
                            <span class={styles.label}>{name}</span>
                        </div>
                    </div>
                </div>
                <div class={ModalStyles.footer}>
                    <button type="reset" value="Reset">
                        Back
                    </button>
                    <button type="submit" value="Submit" disabled={this.state.status !== 'success'}>
                        Next
                    </button>
                </div>
            </form>
        );
    }
}
