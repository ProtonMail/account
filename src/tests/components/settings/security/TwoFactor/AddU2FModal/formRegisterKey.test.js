import render from 'preact-render-to-string';
import { deep } from 'preact-render-spy';

import { FormRegisterKey } from '../../../../../../components/settings/security/TwoFactor/AddU2FModal/formRegisterKey';
import { ERROR_CODE } from '../../../../../../helpers/u2f';

describe('AddU2FModal step FormRegisterKey', () => {
    test('initial display', () => {
        expect(render(<FormRegisterKey settings={{ addU2FKey: {} }}/>)).toMatchSnapshot();

        const settings = { addU2FKey: { response: { name: 'Test name' }, status: undefined, error: undefined } };
        expect(render(<FormRegisterKey settings={settings}/>)).toMatchSnapshot();
    });

    test('failure display with U2F error', () => {
        const settings = {
            addU2FKey: {
                response: { name: 'Test name' },
                status: 'failure',
                error: { metaData: { code: ERROR_CODE.OTHER_ERROR } }
            }
        };
        expect(render(<FormRegisterKey settings={settings}/>)).toMatchSnapshot();
    });


    test('failure display with non-U2F error', () => {
        const onReset = jest.fn();
        const settings = {
            addU2FKey: {
                response: { name: 'Test name' },
                status: 'failure',
                error: new Error('Random error?./')
            }
        };
        render(<FormRegisterKey settings={settings} onReset={onReset}/>);
        expect(onReset).toBeCalledWith(settings.addU2FKey.error.message);
    });

    test('addU2FAction is correctly called', () => {
        const addU2FKeyRegisterAction = jest.fn();
        const settings = {
            addU2FKey: {
                response: { name: 'Test name' },
                status: 'failure',
                error: { metaData: { code: ERROR_CODE.OTHER_ERROR } }
            }
        };
        const context = deep(<FormRegisterKey
            settings={settings}
            addU2FKeyRegisterAction={addU2FKeyRegisterAction}
        />);
        expect(addU2FKeyRegisterAction).toHaveBeenCalledTimes(1);

        context.find('a').first().simulate('click');
        expect(addU2FKeyRegisterAction).toHaveBeenCalledTimes(2);
    });

    test('submit and cancel', () => {
        const addU2FKeyRegisterAction = jest.fn();

        const onSubmit = jest.fn();
        const onCancel = jest.fn();

        const event = { preventDefault: () => undefined };

        const context = deep(<FormRegisterKey
            settings={{ addU2FKey: {} }}
            onSubmit={onSubmit}
            onCancel={onCancel}
            addU2FKeyRegisterAction={addU2FKeyRegisterAction}
        />);

        context.find('form').simulate('submit', event);
        expect(onSubmit).toHaveBeenCalled();
        expect(onCancel).not.toHaveBeenCalled();

        context.find('form').simulate('reset', event);
        expect(onSubmit).toHaveBeenCalledTimes(1); // the submit event
        expect(onCancel).toHaveBeenCalled(); // only once

        expect(addU2FKeyRegisterAction).toHaveBeenCalledTimes(1);
    });

    test('updating status', () => {
        const addU2FKeyRegisterAction = jest.fn();
        const forbidClosure = jest.fn();

        const context = deep(<FormRegisterKey
            settings={{ addU2FKey: {} }}
            addU2FKeyRegisterAction={addU2FKeyRegisterAction}
            forbidClosure={forbidClosure}
        />);

        expect(context).toMatchSnapshot();
        expect(forbidClosure).toHaveBeenCalledTimes(0);

        context.render(<FormRegisterKey
            settings={{ addU2FKey: { status: 'pending' } }}
            addU2FKeyRegisterAction={addU2FKeyRegisterAction}
            forbidClosure={forbidClosure}
        />);
        expect(context).toMatchSnapshot();
        expect(forbidClosure).toHaveBeenCalledTimes(0);

        context.render(<FormRegisterKey
            settings={{ addU2FKey: { status: 'finished' } }}
            addU2FKeyRegisterAction={addU2FKeyRegisterAction}
            forbidClosure={forbidClosure}
        />);
        expect(context).toMatchSnapshot();
        expect(forbidClosure).toHaveBeenCalledTimes(1);


        context.render(<FormRegisterKey
            settings={{ addU2FKey: { status: 'finished' } }}
            addU2FKeyRegisterAction={addU2FKeyRegisterAction}
            forbidClosure={forbidClosure}
        />);
        // second time with the same status, we don't expect forbidClosure is recalled
        expect(forbidClosure).toHaveBeenCalledTimes(1);
    });
});
