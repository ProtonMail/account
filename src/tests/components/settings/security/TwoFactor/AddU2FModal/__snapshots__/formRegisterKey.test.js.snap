// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddU2FModal step FormRegisterKey failure display with U2F error 1`] = `
<form class="wrapper">
  <div class="content container">
    <img src="test-file-stub" />
    <div class="status">
      <span>An error occurred</span>
      <button type="button" class="button">Retry</button>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit" disabled>Next</button>
  </div>
</form>
`;

exports[`AddU2FModal step FormRegisterKey initial display 1`] = `
<form class="wrapper">
  <div class="content container">
    <img src="test-file-stub" />
    <div class="status">
      <div class="row">
        <span class="text">Activate your key</span>
        <span class="text">fetching...</span>
      </div>
      <div class="row">
        <span class="text">Name</span>
        <span class="text nameLabel"></span>
      </div>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit" disabled>Next</button>
  </div>
</form>
`;

exports[`AddU2FModal step FormRegisterKey initial display 2`] = `
<form class="wrapper">
  <div class="content container">
    <img src="test-file-stub" />
    <div class="status">
      <div class="row">
        <span class="text">Activate your key</span>
        <span class="text">fetching...</span>
      </div>
      <div class="row">
        <span class="text">Name</span>
        <span class="text nameLabel">Test name</span>
      </div>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit" disabled>Next</button>
  </div>
</form>
`;

exports[`AddU2FModal step FormRegisterKey updating status 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content container">
    <img src="test-file-stub" />
    <div class="status">
      <div class="row">
        <span class="text">Activate your key</span>
        <span class="text">fetching...</span>
      </div>
      <div class="row">
        <span class="text">Name</span>
        <span class="text nameLabel"></span>
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={false}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={true}
    >
      Next
    </button>
  </div>
</form>

`;

exports[`AddU2FModal step FormRegisterKey updating status 2`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content container">
    <img src="test-file-stub" />
    <div class="status">
      <div class="row">
        <span class="text">Activate your key</span>
        <span class="text">pending...</span>
      </div>
      <div class="row">
        <span class="text">Name</span>
        <span class="text nameLabel"></span>
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={false}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={true}
    >
      Next
    </button>
  </div>
</form>

`;

exports[`AddU2FModal step FormRegisterKey updating status 3`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content container">
    <img src="test-file-stub" />
    <div class="status">
      <div class="row">
        <span class="text">Activate your key</span>
        <span class="text">finished...</span>
      </div>
      <div class="row">
        <span class="text">Name</span>
        <span class="text nameLabel"></span>
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={true}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={false}
    >
      Next
    </button>
  </div>
</form>

`;
